src/spxmod/regmod_builder.py

```diff
@@ -0,0 +1,246 @@
+import numpy as np
+import scipy.sparse as sp
+from msca.linalg.matrix import Matrix, asmatrix
+from msca.optim.solver import IPSolver, NTSolver
+from regmod.data import Data
+from regmod.models import BinomialModel, GaussianModel, PoissonModel
+from regmod.prior import GaussianPrior, LinearGaussianPrior, UniformPrior
+from regmod.variable import Variable
+from scipy.stats import norm
+
+from spxmod.linalg import get_pred_var
+from spxmod.typing import Callable, DataFrame, NDArray, RegmodModel
+
+
+def msca_optimize(
+    model: RegmodModel, x0: NDArray | None = None, options: dict | None = None
+) -> NDArray:
+    x0 = np.zeros(model.size) if x0 is None else x0
+    options = options or {}
+
+    if model.cmat.size == 0:
+        solver = NTSolver(model.objective, model.gradient, model.hessian)
+    else:
+        solver = IPSolver(
+            model.objective,
+            model.gradient,
+            model.hessian,
+            model.cmat,
+            model.cvec,
+        )
+    result = solver.minimize(x0=x0, **options)
+    model.opt_result = result
+    model.opt_coefs = result.x.copy()
+    model.opt_hessian = model.hessian(model.opt_coefs)
+    model.opt_jacobian2 = model.jacobian2(model.opt_coefs)
+    return result.x
```

Comment:
We'll need some way to pass through data about the linear solver to use, or pass through an optional auxiliary function like lin_solve to do specialized solves.

Reply:
Yes! currently this is done through the `options` argument, and we can pass the following argument to control it
https://github.com/ihmeuw-msca/msca/blob/main/src/msca/optim/solver/ntsolver.py#L118-L119